src/armscan_env/volumes/slicing.py

```diff
@@ -0,0 +1,211 @@
+    )
+
+    # Verify the shapes
+    log.debug(
+        f"Original Array Shape: {original_array.shape}\n"
+        f"Padded Array Shape: {padded_array.shape}",
+    )
+
+    return padded_array
+
+
+class EulerTransform:
+    def __init__(self, action: ManipulatorAction, origin: np.ndarray = np.zeros(3)):
+        self.action = action
+        self.origin = origin
+
+    def get_transform(self) -> np.ndarray:
+        # Euler's transformation
+        # Rotation is defined by three rotations around z1, x2, z2 axis
+        th_z1 = np.deg2rad(self.action.rotation[0])
+        th_x2 = np.deg2rad(self.action.rotation[1])
+
+        # transformation simplified at z2=0 since this rotation is never performed
+        return np.array(
+            [
+                [
+                    np.cos(th_z1),
+                    -np.sin(th_z1) * np.cos(th_x2),
+                    np.sin(th_z1) * np.sin(th_x2),
+                    self.origin[0] + self.action.translation[0],
+                ],
+                [
+                    np.sin(th_z1),
+                    np.cos(th_z1) * np.cos(th_x2),
+                    -np.cos(th_z1) * np.sin(th_x2),
+                    self.origin[1] + self.action.translation[1],
+                ],
+                [0, np.sin(th_x2), np.cos(th_x2), self.origin[2]],
+                [0, 0, 0, 1],
+            ],
+        )
+
+    @staticmethod
+    def get_angles_from_rotation_matrix(rotation_matrix: np.ndarray) -> np.ndarray:
+        """Get the angles from a rotation matrix."""
+        # Extract the angles from the rotation matrix
+        th_x2 = np.arcsin(rotation_matrix[2, 1])
+        th_z1 = np.arcsin(rotation_matrix[1, 0])
+
+        # Convert the angles to degrees
+        th_z1 = np.rad2deg(th_z1)
+        th_x2 = np.rad2deg(th_x2)
+
+        return np.array([th_z1, th_x2])
+
+    def transform_action(self, relative_action: ManipulatorAction) -> ManipulatorAction:
+        """Transform an action to be relative to the new coordinate system."""
+        volume_transform_matrix = self.get_transform()
+
+        action_matrix = EulerTransform(relative_action).get_transform()
+        new_action_matrix = np.dot(
+            np.linalg.inv(volume_transform_matrix),
+            action_matrix,
+        )  # 1_A_s = 1_T_0 * 0_A_s
+
+        # new_action_translation = action_matrix[:2, 3] - volume_transform_matrix[:2, 3]
+        new_action_rotation = self.get_angles_from_rotation_matrix(new_action_matrix[:3, :3])
+        new_action_translation = new_action_matrix[:2, 3]
+
+        transformed_action = ManipulatorAction(
+            rotation=new_action_rotation,
+            translation=new_action_translation,
+        )
+
+        log.debug(
+            f"Random transformation: {self.action}\n"
+            f"Original action: {relative_action}\n"
+            f"Transformed action: {transformed_action}\n",
+        )
+
+        return transformed_action
+
+
+class TransformedVolume(sitk.Image):
```

Comment:
this kind of inheritance can be very problematic. I will push a change (still hacky, but better than this)

Reply:
I stuck to the current interface in the end and just added some "safety" features